123-job-jwt/proposal.md

```diff
@@ -0,0 +1,73 @@
+# Summary
+
+Mint a signed JSON Web Token (JWT) and deliver to the job runtime container.
+Additionally, provide a validation endpoint to allow for external or internal
+validation of the issued JWT.
+
+## Motivation
+
+Bluntly, the native integration with Hashicorp Vault is scoped only to static
+secrets and this implementation is inflexible. By leveraging JWT auth at a job
+level, the claims can become incredibly granular, and allow for flexible policy
+assignment. Additionally, this type of framework would allow authentication to
+any number of other frameworks that support the same standard. Then, a more
+complete Vault integration can be built on top of this, as could any integration
+that supports the JWT standard. Lastly, it removes the need for Concourse to
+connect to only one Vault instance, only one Vault namespace, only one Vault
+engine, and have Concourse show up as a single entity in Vault audit logs with a
+policy that is far from being least privilege.
+
+With a JWT Vault Authentication model, each team could have their pipelines
+connect to their own isolated instance or namespace with policy sets tailored to
+the needs of the pipeline jobs, which would enable least privliege models. This
+would also scale better from the Vault perspective, and instead of Concourse
+having to deliver secrets securely from a global persona, a simple trust
+relationship can be established to the Concourse JWT validation API endpoint.
+
+Enabling JWT even opens up the ability to allow Concourse pipelines to
+authenticate and authorize to arbitrary custom APIs if desired.
+
+## Proposal
+
+Adopt [RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519) and a Go library
+that [implements](https://github.com/golang-jwt/jwt) it. Additionally adopt
+[RFC 7517](https://datatracker.ietf.org/doc/html/rfc7517) and a Go library that
+implements it. (Perhaps [jwx](https://github.com/lestrrat-go/jwx)?)
+
+Create a struct that includes as many unique job metadatum as possible,
+including standard claims of `iss`, `sub`, `aud`, `exp`, `jti`, `nbf`, `iat`,
+and Concourse standard objects such as `team`, `teamid`, `job`, `jobid`,
+`pipeline`, `pipelineid`, `task`, `buildid`, and any other data points that
+describe the job and could drive authentication and authorization claims.
+
+Create a public API endpoint that responds to requests where the signed JWT is
+part of the payload, likely leveraging JWKS.
+
+## Open Questions
```

Comment:
How does this work with `var_sources`? https://concourse-ci.org/vars.html#var-sources

Reply:
It could work in parallel, actually.
In the Hashicorp Vault context:
- Pipeline declares in var_sources block:
  - `VAULT_ADDR`
  - `VAULT_AUTH_PATH`
  - `VAULT_AUTH_TYPE` = "jwt"
  - `VAULT_AUTH_ROLE`
  - optionally, `VAULT_NAMESPACE` (default: "/root")
- var_sources constructs an API call:
  - payload:
```jwt
{
  "role": "${VAULT_AUTH_ROLE}",
  "jwt": "${CONCOURSE_CI_JWT}"
}
```
  - `curl -X POST -H "X-Vault-Namespace: ${VAULT_NAMESPACE}" --data @payload.json ${VAULT_ADDR}/v1/auth/${VAULT_AUTH_PATH}/login`
- Concourse stores response:
  - `VAULT_TOKEN` from `.auth.client_token`
  - `VAULT_ACCESSOR` from `.auth.accessor` (for troubleshooting, safe to print to logs)
- Job runs, gets secrets from Vault using the token as necessary
- By default, as cleanup, Token should be revoked, but troubleshooting override for this step should be available

Absolutely can be configured to be leveraged by the `var_sources` block, but best of all, nowhere in this pattern is a human configuring, specifying, or storing a secret in one of the required parameters.

Reply:
added a section `### Potential Workflow in Hashicorp Vault context` to the markdown - please review and see if this addressed your question @taylorsilva